keps/sig-node/5322-dra-driver-permanent-allocation-failure/README.md

```diff
@@ -0,0 +1,1016 @@
+bogged down.
+-->
+
+#### Efficiency
+
+As a cluster administrator, I want to minimize the amount of unnecessary work
+done by critical components like the kubelet and DRA drivers to maximize their
+availability for more important work.
+
+#### Visibility of Errors
+
+As a workload administrator, I want to ensure that my workloads are able to
+start up as quickly and reliably as possible by proactively rescheduling Pods
+when their allocated DRA resources cannot be fulfilled.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+By retrying in all error cases, the current implementation is resilient to
+transient errors. Adding a path that does not retry opens up the possibility
+for miscategorized errors to cause a Pod to terminally fail even when a
+subsequent attempt might allow the Pod's startup to progress.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Broadly, two changes are required to enable permanent errors to skip being
+retried:
+
+1. Allow DRA drivers to annotate errors encountered during the
+   `NodePrepareResources` gRPC call as being permanent.
+1. Update the kubelet to skip retrying to allocate resources for a Pod when a
+   previous request for that Pod failed permanently.
+
+### API
+
+The kubelet's DRA gRPC interface will be updated to allow drivers to express
+that an error is permanent with a new `permanent_error` field:
+
+```proto
+message NodePrepareResourceResponse {
+    // These are the additional devices that kubelet must
+    // make available via the container runtime. A claim
+    // may have zero or more requests and each request
+    // may have zero or more devices.
+    repeated Device devices = 1;
+    // If non-empty, preparing the ResourceClaim failed.
+    // Devices are ignored in that case.
+    string error = 2;
+    // When true and a non-empty error is returned, indicates that the
+    // error is permanent. Permanent errors are expected to fail
+    // consistently for a given request and should not be retried.
+    bool permanent_error = 3;
+}
+```
+
+### Kubelet Handling
```

Comment:
should this be "pod admission failure" similar to https://github.com/kubernetes/kubernetes/pull/132933/?

Reply:
Is "admission" the right word to use for failures during Pod sandbox creation? I see those volume checks happen before that, so just checking to make sure there isn't some threshold in between that and where DRA resources are prepared where we'd call this something different.

Reply:
Thanks! Updated the heading.